File: scripts/build_ext_lookup.py
#!/usr/bin/env python3
"""Build data/ext_lookup.json.gz for the browser extension lookup endpoint.

Merges international journals.json with the domestic ranking sources
(CSSCI / 北大核心 / 科协 cnkx / CSCD / CCF-T / 浙大 / NSFC) into one flat list
of badge objects. The Worker (worker/src/ext-lookup.js) builds ISSN + name
indexes from this file at cold start, so we only emit raw names/issns here and
let the Worker normalize names with the shared pick-match norm().
"""
import gzip
import json
import re
import unicodedata
from datetime import datetime, timezone
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
DATA = ROOT / "data"
OUT = DATA / "ext_lookup.json.gz"


def norm(s):
    """Mirror js/pick-match.js norm() closely (NFKD, strip marks, alnum+CJK)."""
    if not s:
        return ""
    s = unicodedata.normalize("NFKD", str(s)).lower()
    s = "".join(c for c in s if not unicodedata.combining(c))
    s = s.replace("&", " and ")
    s = re.sub(r"[^a-z0-9一-鿿]+", " ", s)
    s = re.sub(r"\s+", " ", s).strip()
    return s


def issn_key(v):
    if not v:
        return ""
    return re.sub(r"[^0-9Xx]", "", str(v)).upper()


def load(name):
    p = DATA / name
    with open(p, "r", encoding="utf-8") as f:
        return json.load(f)


def load_gzip_optional(name):
    p = DATA / name
    if not p.exists():
        return {}
    with gzip.open(p, "rt", encoding="utf-8") as f:
        return json.load(f)


def source_generated_at():
    sources = [
        DATA / "journals.json",
        DATA / "domestic.json",
        DATA / "cscd_journals.json",
        DATA / "fms_journals.json",
        DATA / "vhb_journals.json",
        DATA / "cnrs_journals.json",
        DATA / "scd_journals.json",
        DATA / "ami_journals.json",
        DATA / "annual_outputs.json.gz",
        DATA / "retraction_metrics.json.gz",
    ]
    latest = max(p.stat().st_mtime for p in sources if p.exists())
    return datetime.fromtimestamp(latest, timezone.utc).isoformat()


def records(obj):
    if isinstance(obj, list):
        return obj
    if isinstance(obj, dict) and isinstance(obj.get("records"), list):
        return obj["records"]
    return []


def main():
    journals = load("journals.json")
    dom = load("domestic.json")
    cscd = records(load("cscd_journals.json"))
    retraction_metrics = load_gzip_optional("retraction_metrics.json.gz").get("metrics", {})

    # ── badge objects, keyed for merging ──
    # canonical key: prefer issn, else normalized name
    by_issn = {}   # issnKey -> badge
    by_name = {}   # normName -> badge
    out = []

    def new_badge(name="", issn="", eissn=""):
        b = {"name": name or "", "issn": issn or "", "eissn": eissn or ""}
        out.append(b)
        if issn_key(issn):
            by_issn.setdefault(issn_key(issn), b)
        if issn_key(eissn):
            by_issn.setdefault(issn_key(eissn), b)
        nk = norm(name)
        if nk:
            by_name.setdefault(nk, b)
        return b

    def find_or_create(name="", issn="", eissn="", cn_code=""):
        for k in (issn_key(issn), issn_key(eissn)):
            if k and k in by_issn:
                b = by_issn[k]
                nk = norm(name)
                if nk:
                    by_name.setdefault(nk, b)
                    if re.search(r"[\u4e00-\u9fff]", str(name)) and not b.get("cn_name"):
                        b["cn_name"] = name
                return b
        nk = norm(name)
        if nk and nk in by_name:
            b = by_name[nk]
            if issn and not b.get("issn"):
                b["issn"] = issn
                by_issn.setdefault(issn_key(issn), b)
            if eissn and not b.get("eissn"):
                b["eissn"] = eissn
                by_issn.setdefault(issn_key(eissn), b)
            return b
        return new_badge(name=name, issn=issn, eissn=eissn)

    def retraction_for(issn="", eissn="", name=""):
        for key in (issn, eissn, norm(name)):
            if key and key in retraction_metrics:
                return retraction_metrics[key]
        return None

    # 1) international backbone
    for j in journals:
        b = new_badge(j.get("name", ""), j.get("issn", ""), j.get("eissn", ""))
        if j.get("cn_name"):
            b["cn_name"] = j["cn_name"]
            nk = norm(j["cn_name"])
            if nk:
                by_name.setdefault(nk, b)
        b["slug"] = j.get("slug", "")
        if j.get("if_2024") not in (None, ""):
            b["if_2024"] = j["if_2024"]
        if j.get("if_quartile"):
            b["if_quartile"] = j["if_quartile"]
        if j.get("cas_zone"):
            b["cas_zone"] = j["cas_zone"]
        if j.get("cas_top"):
            b["cas_top"] = True
        xr = j.get("cas_xr")
        if isinstance(xr, dict) and xr.get("zone"):
            b["cas_xr"] = {"zone": xr.get("zone"), "top": bool(xr.get("top"))}
        if j.get("ccf"):
            b["ccf"] = j["ccf"]
        if j.get("abdc"):
            b["abdc"] = j["abdc"]
        if j.get("abs"):
            b["abs"] = j["abs"]
        if j.get("indices"):
            b["indices"] = j["indices"]
        if j.get("scopus"):
            sc = j.get("scopus")
            b["scopus"] = True
            if isinstance(sc, dict):
                b["scopus_active"] = sc.get("active") is not False
                for src, dst in (
                    ("coverage", "scopus_coverage"),
                    ("discontinued", "scopus_discontinued"),
                    ("added_to_list", "scopus_added_to_list"),
                ):
                    if sc.get(src) not in (None, ""):
                        b[dst] = sc.get(src)
        if j.get("free"):
            b["free"] = True
        if j.get("cas_mega"):
            b["cas_mega"] = True
        if j.get("fms"):
            fms = j.get("fms")
            if isinstance(fms, dict) and fms.get("tier"):
                b["fms"] = {"tier": fms.get("tier"), "year": fms.get("year"), "type": fms.get("type")}
        if j.get("vhb"):
            vhb = j.get("vhb")
            if isinstance(vhb, list):
                b["vhb"] = [
                    {
                        "area_code": x.get("area_code"),
                        "rating": x.get("rating"),
                        "year": x.get("year"),
                        "votes": x.get("votes_ge_rating_percent"),
                    }
                    for x in vhb
                    if isinstance(x, dict) and x.get("rating")
                ]
        if j.get("cnrs"):
            cnrs = j.get("cnrs")
            if isinstance(cnrs, list):
                b["cnrs"] = [
                    {
                        "domain": x.get("domain"),
                        "category": x.get("category"),
                        "year": x.get("year"),
                        "historical": True,
                    }
                    for x in cnrs
                    if isinstance(x, dict) and x.get("category")
                ]
        if j.get("scd"):
            scd = j.get("scd")
            if isinstance(scd, dict):
                b["scd"] = {
                    "year": scd.get("year"),
                    "category": scd.get("category"),
                    "newly_added": bool(scd.get("newly_added")),
                }
        if j.get("ami"):
            ami = j.get("ami")
            if isinstance(ami, dict):
                b["ami"] = {"tier": ami.get("tier"), "year": ami.get("year"), "discipline": ami.get("discipline")}
        doaj = j.get("doaj")
        if isinstance(doaj, dict) and doaj.get("apc") not in (None, ""):
            b["doaj_apc"] = str(doaj["apc"])
        if j.get("warning"):
            b["warning"] = True
        if j.get("citic_warning"):
            b["citic_warning"] = True
        if j.get("on_hold"):
            b["on_hold"] = True
        if j.get("under_review"):
            b["under_review"] = True
        rm = retraction_for(j.get("issn", ""), j.get("eissn", ""), j.get("name", ""))
        if rm:
            compact = {
                "total": rm.get("retractions_total"),
                "r5": rm.get("retractions_5y"),
                "r10": rm.get("retractions_10y"),
                "rate5": rm.get("rate_per_1000_5y"),
                "rate10": rm.get("rate_per_1000_10y"),
            }
            b["retraction"] = {k: v for k, v in compact.items() if v not in (None, "", 0)}

    # 2) domestic enrichments (match by issn → name, else create)
    for r in records(dom.get("cssci_core")):
        if not r.get("name") or r["name"].startswith("序号"):
            continue
        find_or_create(name=r["name"])["cssci"] = "core"
    for r in records(dom.get("cssci_ext")):
        if not r.get("name") or r["name"].startswith("序号"):
            continue
        b = find_or_create(name=r["name"])
        b.setdefault("cssci", "ext")
    for r in records(dom.get("pku_core")):
        if not r.get("name"):
            continue
        find_or_create(name=r["name"])["pku"] = True
    for r in records(dom.get("zju")):
        nm = (r.get("name") or "").rstrip("*").strip()
        if not nm:
            continue
        b = find_or_create(name=nm, issn=r.get("issn", ""))
        b["zju"] = r.get("tier", "")
    for r in records(dom.get("nsfc_mgmt")):
        if not r.get("name"):
            continue
        find_or_create(name=r["name"])["nsfc_mgmt"] = r.get("tier", "")
    for r in records(dom.get("ccft")):
        nm = r.get("cn_name") or r.get("en_name") or ""
        if not nm:
            continue
        b = find_or_create(name=nm)
        b["ccft"] = r.get("tier", "")
        if r.get("en_name"):
            b.setdefault("name", r["en_name"])
    for r in cscd:
        if not r.get("name"):
            continue
        b = find_or_create(name=r["name"], issn=r.get("issn", ""))
        b["cscd"] = r.get("database_label") or r.get("database") or "CSCD"

    for r in records(dom.get("cstpcd")):
        if not r.get("name"):
            continue
        b = find_or_create(name=r["name"], issn=r.get("issn", ""))
        b["cstpcd"] = True

    for r in records(dom.get("scd")):
        if not r.get("name"):
            continue
        b = find_or_create(name=r["name"], issn=r.get("issn", ""))
        b["scd"] = {
            "year": dom.get("scd", {}).get("year") if isinstance(dom.get("scd"), dict) else 2026,
            "category": r.get("category", ""),
            "newly_added": bool(r.get("newly_added")),
        }

    for r in records(dom.get("ami")):
        if not r.get("name"):
            continue
        b = find_or_create(name=r["name"])
        b["ami"] = {"tier": r.get("tier", ""), "year": r.get("year", ""), "discipline": r.get("discipline", "")}

    for r in records(dom.get("vhb")):
        if not r.get("title"):
            continue
        b = find_or_create(name=r["title"], issn=r.get("issn", ""))
        arr = b.setdefault("vhb", [])
        item = {
            "area_code": r.get("area_code", ""),
            "rating": r.get("rating", ""),
            "year": dom.get("vhb", {}).get("year", 2024) if isinstance(dom.get("vhb"), dict) else 2024,
            "votes": r.get("votes_ge_rating_percent"),
        }
        if item["rating"] and not any((x.get("area_code"), x.get("rating")) == (item["area_code"], item["rating"]) for x in arr):
            arr.append(item)

    for r in records(dom.get("cnrs")):
        if not r.get("title"):
            continue
        b = find_or_create(name=r["title"], issn=r.get("issn", ""))
        arr = b.setdefault("cnrs", [])
        item = {
            "domain": r.get("domain", ""),
            "category": r.get("category", ""),
            "year": r.get("year", 2020),
            "historical": True,
        }
        if item["category"] and not any((x.get("domain"), x.get("category")) == (item["domain"], item["category"]) for x in arr):
            arr.append(item)

    # 科协 cnkx: a journal can span multiple disciplines → list of {tier,domain}
    for r in records(dom.get("cnkx")):
        tier = r.get("tier", "")
        if not r.get("name") or not re.match(r"^T[123]$", tier or ""):
            continue
        b = find_or_create(name=r["name"], issn=r.get("issn", ""), eissn=r.get("eissn", ""))
        arr = b.setdefault("cnkx", [])
        key = (tier, r.get("domain", ""))
        if not any((x.get("tier"), x.get("domain")) == key for x in arr):
            arr.append({"tier": tier, "domain": r.get("domain", "")})

    payload = {
        "generated_at": source_generated_at(),
        "count": len(out),
        "journals": out,
    }
    raw = json.dumps(payload, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
    with open(OUT, "wb") as raw_out:
        with gzip.GzipFile(filename="", mode="wb", fileobj=raw_out, mtime=0) as f:
            f.write(raw)
    print(f"wrote {OUT} — {len(out)} journals, {OUT.stat().st_size/1024/1024:.1f} MB gz")
    # quick stats
    def cnt(k):
        return sum(1 for b in out if k in b)
    for k in ("cas_zone", "cas_mega", "if_quartile", "fms", "vhb", "cnrs", "scd", "ami", "retraction", "cssci", "pku", "cnkx", "cscd", "ccft", "zju", "indices"):
        print(f"  with {k}: {cnt(k)}")


if __name__ == "__main__":
    main()

File: scripts/build_ext_lookup.mjs
import fs from 'fs';
import zlib from 'zlib';

const DATA_DIR = new URL('../data/', import.meta.url);
const OUT_FILE = new URL('../data/ext_lookup.json.gz', import.meta.url);
// Use a versioned URL for Worker fetches so the immutable Pages cache cannot
// return the previous 4 MB payload after a data refresh.
const VERSIONED_OUT_FILE = new URL('../data/ext_lookup_v3.json.gz', import.meta.url);
const SHARD_DIR = new URL('../data/ext_lookup_v3_shards/', import.meta.url);
const SHARD_COUNT = 64;

function readJson(name, gz = false) {
  const file = new URL(name, DATA_DIR);
  const raw = fs.readFileSync(file);
  const text = gz ? zlib.gunzipSync(raw).toString('utf8') : raw.toString('utf8');
  return JSON.parse(text);
}

function norm(value) {
  return String(value || '')
    .toLowerCase()
    .normalize('NFKD')
    .replace(/[\u0300-\u036f]/g, '')
    .replace(/&/g, ' and ')
    .replace(/[^a-z0-9\u4e00-\u9fff]+/g, ' ')
    .replace(/\s+/g, ' ')
    .trim()
    .replace(/^(the|a|an) /, '');
}

function issnKey(value) {
  return String(value || '').replace(/[^0-9Xx]/g, '').toUpperCase();
}

function shardId(value) {
  let hash = 2166136261;
  const text = String(value || '');
  for (let i = 0; i < text.length; i += 1) {
    hash ^= text.charCodeAt(i);
    hash = Math.imul(hash, 16777619);
  }
  return String((hash >>> 0) % SHARD_COUNT).padStart(2, '0');
}

function cleanCnCode(value) {
  return String(value || '').replace(/^CN/i, '').trim().toUpperCase();
}

function compactJournal(j) {
  const out = {};
  const fields = [
    'name', 'cn_name', 'issn', 'eissn', 'slug', 'publisher',
    'indices', 'scopus', 'cstpcd', 'cscd', 'cssci', 'pku',
    'if_quartile', 'cas_zone', 'cas_top', 'if_2024', 'if_latest',
    'cas_xr', 'ccf', 'abdc', 'abs', 'free',
    'inspec', 'fsta', 'fsta_full_text', 'cabi',
    'warning', 'citic_warning', 'on_hold', 'under_review', 'retraction',
  ];
  fields.forEach((key) => {
    if (j[key] !== undefined && j[key] !== null && j[key] !== '') out[key] = j[key];
  });
  if (j.doaj && typeof j.doaj === 'object') {
    out.doaj = {
      apc: j.doaj.apc || '',
      lic: j.doaj.lic || '',
    };
  }
  return out;
}

const records = [];
const byName = new Map();
const byIssn = new Map();
const byCode = new Map();

function remember(record) {
  const names = [record.name, record.cn_name, ...(record.aliases || [])].filter(Boolean);
  names.forEach((name) => {
    const key = norm(name);
    if (key) byName.set(key, record);
  });
  [record.issn, record.eissn].filter(Boolean).forEach((issn) => byIssn.set(String(issn).toUpperCase(), record));
  if (record.cn_code) byCode.set(cleanCnCode(record.cn_code), record);
}

function findRecord(seed = {}) {
  for (const value of [seed.issn, seed.eissn].filter(Boolean)) {
    const hit = byIssn.get(String(value).toUpperCase());
    if (hit) return hit;
  }
  if (seed.cn_code) {
    const hit = byCode.get(cleanCnCode(seed.cn_code));
    if (hit) return hit;
  }
  for (const value of [seed.name, seed.cn_name, seed.en_name, seed.title, seed.journal_title].filter(Boolean)) {
    const hit = byName.get(norm(value));
    if (hit) return hit;
  }
  return null;
}

function ensureRecord(seed = {}) {
  const hit = findRecord(seed);
  if (hit) return hit;
  const record = {};
  record.name = seed.name || seed.cn_name || seed.en_name || seed.title || seed.journal_title || '';
  if (seed.cn_name && seed.cn_name !== record.name) record.cn_name = seed.cn_name;
  if (seed.en_name && seed.en_name !== record.name) {
    record.aliases = [...new Set([...(record.aliases || []), seed.en_name])];
  }
  if (seed.issn) record.issn = seed.issn;
  if (seed.eissn) record.eissn = seed.eissn;
  if (seed.cn_code) record.cn_code = seed.cn_code;
  records.push(record);
  remember(record);
  return record;
}

function pushUniqueArray(record, key, value) {
  if (!value) return;
  const arr = Array.isArray(record[key]) ? record[key] : [];
  const text = String(value);
  if (!arr.includes(text)) arr.push(text);
  record[key] = arr;
}

function mergeAlias(record, value) {
  if (!value || value === record.name || value === record.cn_name) return;
  record.aliases = [...new Set([...(record.aliases || []), value])];
}

function mergeDomestic(seed, patch) {
  const record = ensureRecord(seed);
  if (seed.cn_code && !record.cn_code) record.cn_code = seed.cn_code;
  if (seed.issn && !record.issn) record.issn = seed.issn;
  if (seed.eissn && !record.eissn) record.eissn = seed.eissn;
  if (seed.cn_name && !record.cn_name && seed.cn_name !== record.name) record.cn_name = seed.cn_name;
  Object.assign(record, patch);
  mergeAlias(record, seed.en_name);
  remember(record);
  return record;
}

const journalsRaw = readJson('journals.json.gz', true);
const journals = Array.isArray(journalsRaw) ? journalsRaw : (journalsRaw.journals || journalsRaw.records || []);
for (const journal of journals) {
  const record = compactJournal(journal);
  records.push(record);
  remember(record);
}

const domestic = readJson('domestic.json');

for (const r of (domestic.cnkx?.records || [])) {
  if (!r.name || !/^T[123]$/i.test(String(r.tier || ''))) continue;
  const record = ensureRecord(r);
  if (r.cn_code && !record.cn_code) record.cn_code = r.cn_code;
  if (r.issn && !record.issn) record.issn = r.issn;
  if (r.eissn && !record.eissn) record.eissn = r.eissn;
  const item = { tier: String(r.tier).toUpperCase(), domain: r.domain || '', subdomain: r.subdomain || '' };
  const key = `${item.tier}|${item.domain}|${item.subdomain}`;
  const existing = new Set((record.cnkx || []).map((x) => `${x.tier}|${x.domain || ''}|${x.subdomain || ''}`));
  if (!existing.has(key)) record.cnkx = [...(record.cnkx || []), item];
  remember(record);
}

(domestic.cssci_core || []).forEach((r) => mergeDomestic(r, { cssci: 'core', cssci_discipline: r.discipline || '' }));
(domestic.cssci_ext || []).forEach((r) => {
  if (!r.name || /^序号\s*期刊名称$/.test(r.name)) return;
  mergeDomestic(r, { cssci: 'ext', cssci_discipline: r.discipline || '' });
});
(domestic.pku_core || []).forEach((r) => mergeDomestic(r, { pku: true, pku_category: r.category || '' }));

for (const r of (domestic.cscd?.records || [])) {
  mergeDomestic(r, {
    cscd: String(r.database || '').toUpperCase() === 'E' ? 'E' : (String(r.database || '').toUpperCase() === 'C' ? 'C' : true),
  });
}

for (const r of (domestic.cstpcd?.records || [])) {
  mergeDomestic(r, { cstpcd: r.kind === 'popular_science' ? 'popular' : true });
}

for (const r of (domestic.scd?.records || [])) {
  mergeDomestic(r, { scd: r.newly_added ? 'SCD+' : 'SCD', scd_category: r.category || '' });
}

for (const r of (domestic.ami?.records || [])) {
  mergeDomestic(r, { ami: r.tier || true, ami_discipline: r.discipline || '' });
}

for (const r of (domestic.ccft || [])) {
  mergeDomestic({ name: r.cn_name, en_name: r.en_name, cn_code: r.cn_code }, { ccft: r.tier || true });
}

for (const r of (domestic.zju?.records || [])) {
  mergeDomestic({ ...r, name: String(r.name || '').replace(/\*$/, '') }, { zju: r.tier || true });
}

for (const r of (domestic.nsfc_mgmt?.records || [])) {
  mergeDomestic(r, { nsfc_mgmt: r.tier || true });
}

for (const r of (domestic.cnki_major?.records || [])) {
  const patch = {
    cn_code: r.cn_code || undefined,
    cnki_category: r.category || '',
    cnki_categories: r.categories || [],
    cnki_tags: r.tags || [],
  };
  if (r.compound_if) patch.cnki_compound_if = r.compound_if;
  if (r.comprehensive_if) patch.cnki_comprehensive_if = r.comprehensive_if;
  mergeDomestic(r, patch);
}

// Keep the extension payload intentionally smaller than the full journal
// bundle.  The extension only needs lookup keys, the detail slug, and the
// fields used to render badges; verbose publisher/Scopus/CNKI records make a
// cold Worker spend tens of megabytes parsing JSON before the first badge can
// be returned.
const EXTENSION_FIELDS = new Set([
  'name', 'cn_name', 'issn', 'eissn', 'slug', 'aliases',
  'indices', 'cstpcd', 'cscd', 'cssci', 'pku',
  'if_quartile', 'cas_zone', 'cas_top', 'if_2024', 'if_latest',
  'ccf', 'free', 'inspec', 'fsta', 'fsta_full_text', 'cabi',
  'warning', 'citic_warning', 'on_hold', 'under_review',
  'cnkx', 'scd', 'ami', 'ccft', 'zju', 'nsfc_mgmt',
]);

function compactExtensionRecord(record) {
  const clean = {};
  EXTENSION_FIELDS.forEach((key) => {
    const value = record[key];
    if (value === undefined || value === null || value === '') return;
    if (Array.isArray(value) && !value.length) return;
    clean[key] = value;
  });

  // Only truthiness is used for the Scopus badge; the source's large status
  // object is not needed by the browser extension.
  if (record.scopus) clean.scopus = true;

  if (record.cas_xr && typeof record.cas_xr === 'object') {
    const zone = record.cas_xr.zone || '';
    if (zone) clean.cas_xr = { zone, ...(record.cas_xr.top ? { top: true } : {}) };
  }

  if (record.doaj && typeof record.doaj === 'object' && record.doaj.apc) {
    clean.doaj = { apc: record.doaj.apc };
  }

  if (record.retraction && typeof record.retraction === 'object') {
    const count = Number(
      record.retraction.retractions_total
      ?? record.retraction.total
      ?? record.retraction.count
      ?? 0,
    );
    if (Number.isFinite(count) && count > 0) clean.retraction = count;
  } else if (Number(record.retraction) > 0) {
    clean.retraction = Number(record.retraction);
  }

  for (const key of ['abdc', 'abs']) {
    if (!record[key]) continue;
    if (typeof record[key] === 'string') clean[key] = record[key];
    else if (record[key].rating || record[key].value) clean[key] = record[key].rating || record[key].value;
  }

  if (Array.isArray(record.cnkx)) {
    clean.cnkx = record.cnkx
      .map((item) => ({ tier: item?.tier || '', domain: item?.domain || '' }))
      .filter((item) => item.tier || item.domain);
    if (!clean.cnkx.length) delete clean.cnkx;
  }

  return clean;
}

const output = records
  .filter((r) => r && (r.name || r.cn_name) && (
    r.issn || r.eissn || r.cn_code || r.slug || r.indices || r.cnkx || r.cssci || r.pku || r.cscd || r.cstpcd || r.scd || r.ami || r.ccft || r.zju || r.nsfc_mgmt || r.cnki_category
  ))
  .map(compactExtensionRecord);

const json = JSON.stringify(output);
const compressed = zlib.gzipSync(json, { level: 9 });
fs.writeFileSync(OUT_FILE, compressed);
fs.writeFileSync(VERSIONED_OUT_FILE, compressed);

// A cold Worker should not parse the entire 11 MB JSON payload for a page
// containing only a handful of journals.  Store direct ISSN/name keys in 64
// immutable shards; a normal batch touches only a few shards in parallel.
const shards = Array.from({ length: SHARD_COUNT }, () => ({}));
output.forEach((record) => {
  const keys = new Set();
  [record.issn, record.eissn].forEach((value) => {
    const key = issnKey(value);
    if (key) keys.add(`i:${key}`);
  });
  [record.name, record.cn_name, ...(record.aliases || [])].forEach((value) => {
    const key = norm(value);
    if (key) keys.add(`n:${key}`);
  });
  keys.forEach((key) => {
    shards[Number(shardId(key))][key] = record;
  });
});
fs.mkdirSync(SHARD_DIR, { recursive: true });
const shardSizes = [];
shards.forEach((shard, index) => {
  const shardJson = JSON.stringify(shard);
  const shardCompressed = zlib.gzipSync(shardJson, { level: 9 });
  const file = new URL(`${String(index).padStart(2, '0')}.json.gz`, SHARD_DIR);
  fs.writeFileSync(file, shardCompressed);
  shardSizes.push(shardCompressed.length);
});
console.log(`wrote ${OUT_FILE.pathname}, ${VERSIONED_OUT_FILE.pathname}, and ${SHARD_COUNT} shards (${output.length.toLocaleString()} records, ${Buffer.byteLength(json).toLocaleString()} bytes raw, ${shardSizes.reduce((a, b) => a + b, 0).toLocaleString()} shard bytes)`);
